src/obfuscator.ts

```diff
@@ -0,0 +1,268 @@
+    projection[f] = 1;
+  })
+  return projection;
+}
+
+const obfuscateableFieldsFromCollection = (collection: Tyr.ObfuscateableCollectionInstance): Array<string> => {
+  const names: Array<string> = [];
+  for (const fieldName in collection.def.fields) {
+    if (collection.def.fields[fieldName].def.obfuscateable) {
+      names.push(fieldName);
+    }
+  }
+  return names;
+}
+const metaDataNameForCollection = (tyrCollectDbName: string):string => {
+  return `${tyrCollectDbName}__meta`
+}
+const getMetaDataCollection = async (name: string): Promise<Collection> => {
+  const existingCollection = await Tyr.db.collection(name);
+  if (!existingCollection) {
+    return createMetaDataCollection(name);
+  }
+  return existingCollection;
+}
+
+//TODO: probably remove, up to consumer to provide collection
+const createMetaDataCollection = async (name: string): Promise<Collection> => {
+  return await Tyr.db.createCollection<Tyr.ObfuscateMetaDataSchema>(name, { autoIndexId: true });
+}
+
+const createBatchMetaData = async (fields: Array<string>, srcCollection: Collection,
+  targetCollection: Collection, query: FilterQuery<object>, replaceVals: object, replaceValCollection: Collection): Promise<Tyr.ObfuscateBatchResult> => {
+  
+  const timeStamp: number = new Date().getTime();
+  const batchTag: ObjectID = new ObjectID();
+  const docIdsCursor: Cursor = await srcCollection.find(query).project(idOnlyProjection);
+
+  const count = await docIdsCursor.count();
+  Tyr.info(`Obfuscating ${count} documents from ${srcCollection.namespace} tag ${batchTag}`);
+
+  let bulkMetaData = targetCollection.initializeUnorderedBulkOp();
+
+  let doc: IdRecord;
+  let metaDataRecord: Tyr.ObfuscateMetaDataSchema;
+  let batchRecIds = [];
+
+  for (let i = 1; i <= count; i++) {
+    doc = await docIdsCursor.next();
+
+    metaDataRecord = {
+      timestamp: timeStamp, recordId: doc._id, fields, batchTag
+    };
+    batchRecIds.push(doc._id);
+    bulkMetaData.insert(metaDataRecord);
+
+    if ((i % defaultBatchSize === 0) || i === count) {
+      try {
+        const metaDataResult: BulkWriteResult = await bulkMetaData.execute();
+      } catch (error) {
+        console.log(`MetaData Error: ${JSON.stringify(error)}`);
+      }
+
+      if (!replaceVals) {
+        //Move data from given collection to be mask
+        applyMaskValuesFromCollection(batchRecIds, replaceValCollection, srcCollection);
+      }
+
+      bulkMetaData = targetCollection.initializeUnorderedBulkOp();
+    }
+  }
+
+  try {
+    //Static replacement values can be applied at once
+    const maskResult: UpdateWriteOpResult = await srcCollection.updateMany(query, { $set: replaceVals });
+  } catch (error) {
+    console.log(`Masking Error: ${JSON.stringify(error)}`);
+  }
+  Tyr.info(`Finished obfuscating collection ${srcCollection.namespace} tag ${batchTag}`);
+
+  // TODO: Get count from batch results
+  return { batchTag: batchTag, count: count };
+}
+
+// TODO: Implement
```

Comment:
Just realized the reason my test was commented out, is because I removed the implementation when I was trying to create a universal 'migrate' function.  Will re-implement